### PR TITLE
feat: Allow for schema evolution

### PR DESCRIPTION
**Pull request by Fokko**
Please don't hold back on my code!

### CODE REVIEW

File: crates/integration_tests/tests/read_evolved_schema.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    let batches: Vec<_> = batch_stream.try_collect().await.unwrap();
+
+    let batch0 = batches[0].column_by_name("language").unwrap();
+    let batch1 = batches[1].column_by_name("language").unwrap();
+
+    assert_eq!(batch0.as_any().downcast_ref::<StringArray>().unwrap().value(0), "Java");
+    assert_eq!(batch1.as_any().downcast_ref::<StringArray>().unwrap().value(0), "Python");
+
+    // Evolve column - Does not yet work
```
Comment (Fokko) [inline crates/integration_tests/tests/read_evolved_schema.rs:57]: With the rule added above, it still fails with:

```
Parquet schema table {
  1: foo: required int 
}
 and Iceberg schema table {
  1: foo: optional long 
}
 do not match.
```

**Comment by Xuanwo** [comment]
Hi @Fokko, I'm following up on issue #790.

It appears this PR doesn't address the scenario I described in #790, where the underlying parquet schema differs from the declared one. Could schema evolution potentially resolve this? I'm concerned we might also need to adjust the Arrow schema mappings instead of just project mask.

**Comment by Fokko** [comment]
This is pretty similar. With this PR we first write an int, and then read it as a long, which is a valid schema evolution. In the case of #790 we try to read a narrower decimal from the Parquet file, into a wider one from the Iceberg schema. I haven't tested it with the decimal, that might be a good followup 👍

**Comment by Xuanwo** [comment]
Thank you for the explanation! This PR is good enough for me. Let's proceed.

**Review by Xuanwo — APPROVED** [review]
Thank you @Fokko for fixing this

**Pull request merged**